tests/test_benchmark.py

```diff
@@ -120,20 +124,153 @@ def process(filename):
     final = final.replace('.cpp', '')
     try_delete(final)
     output = Popen([PYTHON, EMCC, filename, #'-O3',
-                    '-O3', '-s', 'DOUBLE_MODE=0', '-s', 'PRECISE_I64_MATH=0',
+                    '-O3',
                     '--memory-init-file', '0', '--js-transform', 'python hardcode.py',
                     '-s', 'TOTAL_MEMORY=256*1024*1024',
                     '-s', 'NO_EXIT_RUNTIME=1',
+                    '-s', 'NO_FILESYSTEM=1',
+                    '-s', 'EXPORTED_RUNTIME_METHODS=[]',
                     '-s', 'BENCHMARK=%d' % (1 if IGNORE_COMPILATION and not has_output_parser else 0),
                     #'--profiling',
-                    #'--closure', '1',
+                    '--closure', '1',
                     '-o', final] + shared_args + emcc_args + self.extra_args, stdout=PIPE, stderr=PIPE, env=self.env).communicate()
     assert os.path.exists(final), 'Failed to compile file: ' + output[0] + ' (looked for ' + final + ')'
     self.filename = final
 
   def run(self, args):
     return run_js(self.filename, engine=self.engine, args=args, stderr=PIPE, full_output=True, assert_returncode=None)
 
+class CheerpBenchmarker(Benchmarker):
+  def __init__(self, name, engine, args=['-O3'], binaryen_opts=[]):
+    self.name = name
+    self.engine = engine
+    self.args = args
+    self.binaryen_opts = binaryen_opts[:]
+
+  def build(self, parent, filename, args, shared_args, emcc_args, native_args, native_exec, lib_builder, has_output_parser):
+    suffix = filename.split('.')[-1]
+    cheerp_temp = filename + '.cheerp.' + suffix
+    code = open(filename).read()
+    if 'int main()' in code:
+      main_args = ''
+    else:
+      main_args = 'argc, (%(const)s char**)argv' % {
+        'const': 'const' if 'const char *argv' in code else ''
+      }
+    open(cheerp_temp, 'w').write('''
+      %(code)s
+      void webMain() {
+        // TODO: how to read from commandline?
+        volatile int argc = 2;
+        typedef char** charStarStar;
+        volatile charStarStar argv;
+        argv[0] = "./cheerp.exe";
+        argv[1] = "%(arg)s";
+        main(%(main_args)s);
+      }
+    ''' % {
+      'arg': args[-1],
+      'code': code,
+      'main_args': main_args
+    })
+    cheerp_args = [
+      '-target', 'cheerp', '-cheerp-mode=wasm',
+    ]
+    self.parent = parent
+    if lib_builder:
+      # build as "native" (so no emcc env stuff), but with all the cheerp stuff
+      # set in the env
+      cheerp_args = cheerp_args + lib_builder(self.name, native=True, env_init={
+        'CC': '/opt/cheerp/bin/clang',
```

Comment:
This assumes Linux and that one has manually installed cheerp in advance? Btw, did cheerp build their own wasm backend or do they depend on upstream LLVM backend or JSBackend?

Reply:
Yeah, that's how cheerp installs itself on linux, always to that path. Not sure how to get this to work on other platforms.

Cheerp built their [own wasm writing backend](https://github.com/leaningtech/cheerp-llvm/blob/master/lib/CheerpWriter/CheerpWasmWriter.cpp). Mostly straightfoward writing, little wasm-specific optimizations so far (the binaryen optimizer shrinks their binaries by 15%, I mentioned it to them).

Reply:
Is it possible to add a `if (not exists /opt/cheerp/bin/clang) self.skip('missing cheerp')` or similar? I think that's what we added for Crunch and SCons and some others in the suites.

Reply:
sure, adding